crates/bifrost/src/providers/local_loglet/mod.rs

```diff
@@ -230,14 +234,31 @@ impl LogletBase for LocalLoglet {
             Err(ShutdownError.into())
         })?;
 
-        self.last_committed_offset
-            .fetch_max(offset.into(), Ordering::Relaxed);
-        self.notify_readers();
+        // The order here is critical. We must update the last_committed_offset before checking the
+        // seal to ensure that `find_tail` observe `last_committed_offset` changes.
+        let release_pointer = LogletOffset::from(
+            self.last_committed_offset
+                .fetch_max(offset.into(), Ordering::SeqCst)
+                .max(offset.into()),
+        );
+        self.notify_readers(release_pointer);
+        // Ensure that we don't acknowledge the append (even that it has happened) if the loglet
+        // has been sealed already. Memory ordering Acquire is required to establish happens-before
+        // relationship between bumping the last_commit_offset and the seal operation.
+        if self.sealed.load(Ordering::Acquire) {
+            return Err(AppendError::Sealed);
+        }
```

Comment:
I really like the idea of fetch_add(0, Release), very clever solution! 

In all fairness, my approach was to mash something together and slam SeqCst as temporary measure until a follow up PR that will replace the entire module with a sequencer implementation. But since you've done a great job reviewing it, I've updated this version to a nicer/better version. 

In fact, what was making the whole thing work is the Mutex that already acts as a memory barrier/fence which synchronized threads.